File: 155-min-stack/155-min-stack.cpp
class MinStack {
    stack<long long> st;
    long long min_val;
public:
    MinStack() {
        min_val = INT_MAX;
    }
    
    void push(int val) {
        long long value = val;
        if(st.empty()){
            min_val = value;
            st.push(value);
        } else {
            if(value < min_val){
                st.push(2*value - min_val);
                min_val = value;
            } else {
                st.push(value);
            }
        }
    }
    
    void pop() {
        if(st.empty()){
            return;
        }
        long long ele = st.top();
        st.pop();
        if(ele < min_val){
            min_val = 2 * min_val - ele;
        }
    }
    
    int top() {
        if(st.empty()){
            return -1;
        }
        long long ele = st.top();
        if(ele < min_val){
            return min_val;
        }
        return ele;
            
    }
    
    int getMin() {
        return min_val;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.index = 0
        self.min = 0

    def push(self, val: int) -> None:
        if(self.index == 0):
            self.stack.append(val)
            self.min = val
        else:
            if(val<self.min):
                self.stack.append(2 * val - self.min)
                self.min = val
            else:
                self.stack.append(val)
        self.index += 1

    def pop(self) -> None:
        if(self.index <= 0):
            return
        self.index -= 1
        value = self.stack.pop()
        if(value < self.min):
            self.min = 2*self.min - value

    def top(self) -> int:
        if(self.index <= 0):
            return
        value = self.stack[self.index-1]
        if(value < self.min):
            return self.min
        return value

    def getMin(self) -> int:
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()